File: boats.cpp
class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        
        int n = people.size(), p1 = 0, p2 = n - 1, sum = 0, boats = 0;
        
        sort(people.begin(), people.end(), greater<int>());
        
        while(p1 <= p2)
        {
            sum = people[p1] + people[p2];
            
            if(sum <= limit)
            {
                boats++;
                p1++;
                p2--;
            }
            else if(sum > limit)
            {
                boats++;
                p1++;
            }
        }
        
        return boats;
       
    }
};

File: progress_sheet/boats.py
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:

        left = 0
        right = len(people) - 1
        boats = 0

        # sort in increasing order
        people.sort()

        # find the best candidates that can 
        # fit on the boat and count the number of boats used
        while left <= right:

            sum = people[left] + people[right]

            if sum <= limit:
                left += 1
            
            boats += 1
            right -= 1
        
        return boats
